### ISSUE TITLE
Add a lint that warns when binary-crate's main always panics.

### ISSUE THREAD

**Issue by crlf0710** [rust-lang/rust#68245]
This seems a useful lint, but I'm not sure whether this is implementable in `clippy`.

**Comment by crlf0710** [comment]
A common scenario. When a rust beginner learns to write rust code, usually they will experiment in the `main()` function, writing faulty code here and there. Many of these are compile errors, which will be prompted immediately by IDE. For errors in the rest, many of them will lead to an unvoidable runtime panic, rendering the rest of the program dead code.  This already indicates the mistake in the program.

I think it's best if we can inform the user using a lint in this case, and IDEs will show the lints to them. From user's perspective, this functionality is actually a better fit for `clippy`, but i'm not sure if `clippy` is able to gain this information, i guess not.

**Comment by Centril** [comment]
We already have dead / unreachable code lints -- why are those insufficient?

**Comment by crlf0710** [comment]
Today
```rust
fn main() {
    return;
    println!("foo");
}
```
triggers `unreachable_code` lint, while
```rust
fn main() {
    assert!(false);
    println!("foo");
}
```
doesn't trigger any lint by default. I'm suggesting a lint that covers the second case. Do you think the existing lint should be expanded to cover the second case, or there should not be a lint that covers this case?

**Comment by Centril** [comment]
I see; I think the lints would need to be rewritten on MIR to support that. cc @oli-obk @wesleywiser.

**Comment by oli-obk** [comment]
This doesn't need any new lints. The `const_prop` pass should at some point end up triggering on that `assert!(false)`

Amusingly enough we already optimize everything to the point where it should report that. The MIR for `assert!(false)` is 

```rust
_2 = &(promoted[0]: (&str, u32, u32)); // bb0[2]: scope 0 at <::std::macros::panic macros>:4:14: 5:75
_1 = _2;                         // bb0[3]: scope 0 at <::std::macros::panic macros>:4:14: 5:75
const std::rt::begin_panic::<&str>(const "assertion failed: false", move _1);
```

So basically all we need is to tell `const_prop` to run `TerminatorKind::Call` through the miri engine in https://github.com/rust-lang/rust/blob/2480c9eac15608591f58730aed27caac9c30b4c2/src/librustc_mir/transform/const_prop.rs#L1004

This may cause some funny diagnostics though, we need to step carefully and consider what we want to report.
